src/Features/Core/Portable/Completion/Providers/ImportCompletionProvider/AbstractTypeImportCompletionService.cs

```diff
@@ -0,0 +1,381 @@
+            public INamedTypeSymbol BestGenericOverload { get; }
+
+            public bool ContainsPublicGenericOverload { get; }
+
+            public TypeOverloadInfo Aggregate(INamedTypeSymbol type)
+            {
+                if (type.Arity == 0)
+                {
+                    return new TypeOverloadInfo(nonGenericOverload: type, BestGenericOverload, ContainsPublicGenericOverload);
+                }
+
+                // We consider generic with fewer type parameters better symbol to show in description
+                var newBestGenericOverload = BestGenericOverload == null || type.Arity < BestGenericOverload.Arity
+                    ? type
+                    : BestGenericOverload;
+
+                var newContainsPublicGenericOverload = type.DeclaredAccessibility >= Accessibility.Public || ContainsPublicGenericOverload;
+
+                return new TypeOverloadInfo(NonGenericOverload, newBestGenericOverload, newContainsPublicGenericOverload);
+            }
+        }
+
+        private readonly struct ReferenceCacheEntry
+        {
+            public class Builder
+            {
+                private readonly string _language;
+                private readonly string _genericTypeSuffix;
+                private readonly Checksum _checksum;
+
+                public Builder(Checksum checksum, string language, string genericTypeSuffix)
+                {
+                    _checksum = checksum;
+                    _language = language;
+                    _genericTypeSuffix = genericTypeSuffix;
+
+                    _itemsBuilder = ArrayBuilder<TypeImportCompletionItemInfo>.GetInstance();
+                }
+
+                private ArrayBuilder<TypeImportCompletionItemInfo> _itemsBuilder;
+
+                public ReferenceCacheEntry ToReferenceCacheEntry()
+                {
+                    return new ReferenceCacheEntry(
+                        _checksum,
+                        _language,
+                        _itemsBuilder.ToImmutableAndFree());
+                }
+
+                public void AddItem(INamedTypeSymbol symbol, string containingNamespace, bool isPublic)
+                {
+                    var isGeneric = symbol.Arity > 0;
+
+                    // Need to determine if a type is an attribute up front since we want to filter out 
+                    // non-attribute types if in attribute context. We can't do this lazily since we don't hold 
+                    // on to symbols.
+                    //
+                    // Also for attribute types with "Attribute" suffix, we only cache items with their full name 
+                    // (and creating those w/o suffix on ther fly if in attribute context), because we assume it's
+                    // much more common to trigger completion in non-attribute context.
+                    var isAttribute = symbol.IsAttribute();
```

Comment:
This is to support the scenario of completing in attribute context, so we can filter out non-attribute items. However, trace shows that this change single-handedly caused a 600% increase in CPU time (2313ms out of 2732ms spent in import completion provider).

Note that this will mostly only affect first trigger when cache is cold. Given that we are planning to add timeout mechanism  for import completion, this perf regression might be OK. Thoughts?

@CyrusNajmabadi @dpoeschl @ivanbasov 

![image](https://user-images.githubusercontent.com/788783/60373638-b2d5d800-99b5-11e9-825f-f4442b27ca11.png)

Reply:
Can we put in a weak heuristic?  like checking first if the name of hte type ends with 'Attribute'?  Thta's all that would really matter right, since we only use this bit along with that suffix to trim the suffix if we're actually in an attribute context, right?  So why not do the quick string check first?

Reply:
Because we also want to filter out all non-attribute types when triggered in attribute context, to be consistent with regular symbol completion.

Reply:
i'm saying: instead of doing: `var attribute = symbol.Name.EndsWith("Attribute") && symbol.IsAttribute().  It should speed thins up a lot, while only being slightly inaccurate.

Reply:
> while only being slightly inaccurate.

But isn't this a big problem? People would expect the filtering to be conservative.

Reply:
Another question: shouldn't [`IsAttribute`](http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/Shared/Extensions/ITypeSymbolExtensions.cs,13b7fa23c11ae564,references) contain a check for class type?

Reply:
> But isn't this a big problem? People would expect the filtering to be conservative.


Literally 99.9999999% of all attributes end with `Attribute`.  I don't care if we miss 1-2 attributes in the entire world that don't end with `Attribute` but are actually attributes.

Reply:
Hmm, I'm not sure about that... Summoning @dotnet/roslyn-ide for opinions on this.

Reply:
1. It's an easy thing to verify (if you don't trust me).
2. It's so much the pattern for .net that it's *baked into* the language.

i.e. this is the thing that allows you to write `[Foo]` when the name is `FooAttribute`.  It's been the recommended and followed naming across teh ecosystem since, well, forever :)

Reply:
No, I totally agree with you that this is *the* pattern that everyone *should* follow. And anyone who doesn't should not be allowed to write C# code if up to me. However, not having the suffix is not disallowed, and as long as that's the case, I'm sure some people would do just that. 

I guess my reservation about ignoring those attributes is based on my understanding of how Intellisense should work (I might very well be wrong, so please correct me if that's the case), which is it should at least provide every valid type that can be typed here. So it can be used for look up purpose if people want.

Reply:
> However, not having the suffix is not disallowed, and as long as that's the case, I'm sure some people would do just that.

Right.  But:

1. actually look at the data.  I'm fairly certain my number is accurate :)
2. we are not the language/compiler.  We are the IDE.   We can *totally* use a heuristic that is right for 99.999% of cases when it can massively benefit our performance.  

>which is it should at least provide every valid type that can be typed here. So it can be used for look up purpose if people want.

that's a laudable goal :)   however, it's more appropriate to add caveats like "it can be slightly inaccurate if that provides substantial benefits".  For example, it's substantially beneficial to cut down on the CPU.  We've *never* gone for a "100% accurate in all cases" approach for *any* IDE feature.  Instead, we operate on a expectation around diminishing returns (or costs).  

I think it's totally fine to discuss this with the team.  However, i think this is a reasonable sort of heuristic that falls *well* in line with how we generally do things.

Reply:
OK, you have convinced me. Will change. Would you like to do a full review of this PR? :)

Reply:
> Literally 99.9999999% of all attributes end with Attribute

I have longer-term goals to be able to prove things like this, but for now I do also hypothesize that this is a highly followed guideline and that the tradeoff is probably worth it in this case.